### ISSUE TITLE
Media Players Not Updating

### ISSUE THREAD

**Issue by broyuken** [custom-components/alexa_media_player#1953]
**IMPORTANT: Please search the issues, including closed issues, and the [FAQ](https://github.com/custom-components/alexa_media_player/wiki/FAQ) before opening a new issue. The template is mandatory; failure to use it will result in issue closure.**

**Describe the bug**

Media Player entities go stale after a few hours. If I start playing media either with my voice, or by having a routine triggered my HA media player entities don't know about it and just says standby. Or if media has been playing for a while and I stop it, the media player entity will still show it playing. Reloading the integration with kick things and get them showing the correct status again for a little while
**To Reproduce**

1. Play media on an alexa device after it's been idle for a few hours.
2. Observe media state in HA hasn't changed
3. Submit issue on github

**Expected behavior**

Media player should show current status of media player entities
**Screenshots**

**System details**

- Home-assistant (version): 2023.5.3
- alexa_media (version from `const.py` or HA startup): 4.6.4
- alexapy (version from `pip show alexapy` or HA startup): 1.26.8
- Amazon [2FA is enabled](https://github.com/custom-components/alexa_media_player/wiki/Configuration#enable-two-factor-authentication-for-your-amazon-account) (y/n). :yes

**Logs**
Please provide [logs](https://github.com/custom-components/alexa_media_player/wiki/FAQ#how-do-i-enable-debug-logging-for-the-component).
These debug logs begin at 9:08 when I run my "soundmachineon" routine which plays white noise on amazon music and didn't update the status until 9:09 when I reloaded alexa media player. Ignore the BROKEN_TUYA lines, they were filling up my logs so I did a find/replace to get rid of them.

[alexa_logs.txt](https://github.com/custom-components/alexa_media_player/files/11538235/alexa_logs.txt)

**Comment by alandtse** [comment]
The logs indicate that websocket is reporting data, the player urls are being queried, and you're not getting errors to the requests. Please note that not all services actually update the player. Check the official app and see if it's actually displaying what's playing.

**Comment by broyuken** [comment]
Yes, but the logs also show that the routine starts which just plays white noise from Amazon’s music, but the status of the player never updates. The official app does show that update, and reloading the integration updates the media player and then everything is fine…for a couple hours.

**Comment by broyuken** [comment]
I just checked this morning and the Alexa app shows that it is playing, and when I say Alexa stop, it shows it paused. In HA however it still shows it playing. If I reload the component it will update correctly.

**Comment by broyuken** [comment]
Just saw this when I try to pause it in HA when it’s already paused in real life. 

`Logger: homeassistant.components.websocket_api.http.connection
Source: custom_components/alexa_media/media_player.py:1170 
Integration: Home Assistant WebSocket API (documentation, issues) 
First occurred: 6:21:51 AM (2 occurrences) 
Last logged: 6:21:59 AM

[140630914232624] 'NoneType' object has no attribute 'data'
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 199, in handle_call_service
    await hass.services.async_call(
  File "/usr/src/homeassistant/homeassistant/core.py", line 1849, in async_call
    task.result()
  File "/usr/src/homeassistant/homeassistant/core.py", line 1889, in _execute_service
    await cast(Callable[[ServiceCall], Awaitable[None]], handler.job.target)(
  File "/usr/src/homeassistant/homeassistant/helpers/entity_component.py", line 226, in handle_service
    await service.entity_service_call(
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 798, in entity_service_call
    future.result()  # pop exception if have
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 980, in async_request_call
    await coro
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 838, in _handle_entity_call
    await result
  File "/config/custom_components/alexa_media/helpers.py", line 156, in _catch_login_errors
    result = await func(*args, **kwargs)
  File "/config/custom_components/alexa_media/media_player.py", line 1163, in async_media_pause
    await self._playing_parent.async_media_pause()
  File "/config/custom_components/alexa_media/helpers.py", line 156, in _catch_login_errors
    result = await func(*args, **kwargs)
  File "/config/custom_components/alexa_media/media_player.py", line 1170, in async_media_pause
    self.hass.data[DATA_ALEXAMEDIA]["accounts"][self._login.email]["websocket"]
AttributeError: 'NoneType' object has no attribute 'data'
`

**Comment by steff75** [comment]
I have found that it is usually enough to perform a media player action (media_next_track, volume_up, ...) when it hangs. Then the dashboard player status mostly updates immediately. A restart of the integration was not necessary for me so far.

**Comment by broyuken** [comment]
I just called the volume down and volume up services but it still shows it in standby

**Comment by alandtse** [comment]
> Just saw this when I try to pause it in HA when it’s already paused in real life.

This error indicate HA is broken for the integration. Every integration is supposed to have access to the `hass.data` object automatically. If it doesn't, then there is something very wrong. There is no way to solve the issue short of a reload of the integration where hopefully HA provides the entity. Assuming you're not getting that error, then you should confirm you're seeing websocket traffic. You can test this by enabling debug logging and then manually changing the volume button. If you're not seeing websocket, then the integration cannot update info without directly using HA to interact with the component.

**Comment by broyuken** [comment]
I was trying to make sure things were broken before I started the debug logs and I found something interesting. If I call my routine from within HA using a service call from a card like this or from an automation, I can see the volume adjust on the lovelace card, but the state still shows standby.
```
type: custom:mushroom-entity-card
entity: script.music_sound_machine
name: Sound Machine
tap_action:
  action: call-service
  service: script.music_sound_machine
  service_data:
    volume: 0.6
    device: media_player.charlottes_echo_spot
  target: {}
icon: mdi:sleep
secondary_info: none
layout: vertical
fill_container: true
```

However if I go into the alexa app on my phone I can see the media player in a playing state. Then if I hit pause in the alexa app, the lovelace card updates showing it's paused. 

Once it breaks again I will try to enable debug logging and see what happens.

**Comment by broyuken** [comment]
It broke again, and here are some logs. I started my goodnight routine around 2023-06-06 21:02:10.306, but the logs never show the state updating from standby, it always stays <entity media_player.master_bedroom_echo_show=standby>. The alexa app shows the stream playing as an active media source, and if I was to pause it (I didn't because it would wake up my wife) it would trigger HA to notice that it's playing. I tried changing the volume with both service calls and from the HA app, but neither of those triggered the state change.

These are the raw debug logs minus the local tuya stuff I removed from there because it is log spam that I need to fix when I find the time. The other errors you see in there are spotify went funky earlier today and one of my roborocks is offline because it's being replaced:
[home-assistant_alexa_media_2023-06-07T01-06-47.569Z.log](https://github.com/custom-components/alexa_media_player/files/11671565/home-assistant_alexa_media_2023-06-07T01-06-47.569Z.log)

**Comment by alandtse** [comment]
Your websocket is reporting changes but doesn't appear to report anything after your routine is run. If this worked before, it looks like the websocket isn't broadcasting any state changes now so it's an API change. There's nothing I can do to capture this unless after every routine I turn on polling for all alexa devices. That will likely be very spammy and I have historically not enabled polling unless there's actually a play command issued by a media player.

That said, due to the restart, the component does try to make one attempt to see if things are playing. In this case for the device where it appears you issued the request this is the response and it didn't see anything playing per amazon. Are you positive the App or webplayer actually shows white noise content when playing?

```
2023-06-06 21:06:33.897 DEBUG (MainThread) [alexapy.alexaapi] d***********6@y*******m: Trying get: https://alexa.amazon.com/api/np/player?deviceSerialNumber=SNIP&deviceType=A4ZP7ZC4PI6TO&screenWidth=2560&_=1686099993897 : with uri: /api/np/player data None query {'deviceSerialNumber': 'SNIP', 'deviceType': 'A4ZP7ZC4PI6TO', 'screenWidth': 2560, '_': 1686099993897}
2023-06-06 21:06:34.072 DEBUG (MainThread) [alexapy.alexaapi] d***********6@y*******m: GET: https://alexa.amazon.com/api/np/player?deviceSerialNumber=SNIP&deviceType=A4ZP7ZC4PI6TO&screenWidth=2560&_=1686099993897 returned 200:OK:application/json
2023-06-06 21:06:34.074 DEBUG (MainThread) [custom_components.alexa_media.media_player] d***********6@y*******m: Refreshing <entity None=standby>

```

**Comment by broyuken** [comment]
I am 100% positive the Alexa app shows white noise playing.

**Comment by broyuken** [comment]
Are there any other logs I need to provide to help you track this down?

**Comment by broyuken** [comment]
Checking in on this again. Is there anything else you need from me to help move this along?

**Comment by broyuken** [comment]
@alandtse is there anything else I can provide to help solve this issue?

**Comment by alandtse** [comment]
Please don't ping me. I don't promise any support. The code is open and you can make your own changes. I've already analyzed your logs and explained what I know. Honestly, I'm probably done with supporting this and pings do not encourage me to work more.

**Comment by emufan** [comment]
I would completely understand and as you said open code, etc. But I cannot do such codings and have the not-updating-anymore as well. So any mitigation oder fix on the long run would be very very good of course.

Here e.g. die next-timer entities often are not started and stay as unavailable even when a timer on the device is started. Only a reload of the integration is fixing this then.

**Comment by broyuken** [comment]
Sorry for pinging you, the only reason I did was because I sent a few messages with responses to your question and didn’t hear anything back. The app shows it playing, HA does not. 

Is there anything we can do as non coders to help you try to figure out what may have changed in the api? We can’t be the only two people having this issue. 

@emufan couple questions,
1. What type of WiFi APs do you use?
2. What type of echo devices are you seeing this issue on?

for me I use Unifi WiFi 6 Lites and am seeing this issue in particular on an echo show 5 and an echo spot. Those are the two devices I run the sound machine on, but I “think” the issue occurs on them all.

**Comment by emufan** [comment]
AVM Fritz Repeater 2400 and Wchoe Show 5.

**Comment by krisnoble** [comment]
Does running homeassistant.update_entity on the relevant media_player entity help at all?

**Comment by broyuken** [comment]
That’s a great question, I’m not home this weekend but I’ll test that out when I get back.

**Comment by broyuken** [comment]
I have issues with my printer as well where it’s unavailable for AirPrint unless I do a reconnect on it in the unifi app.

**Comment by RSpeekenbrink** [comment]
> Does running homeassistant.update_entity on the relevant media_player entity help at all?

This did it for me, had the same problem, but manually pinging an entity update does the trick.

**Comment by RSpeekenbrink** [comment]
> > Does running homeassistant.update_entity on the relevant media_player entity help at all?
> 
> This did it for me, had the same problem, but manually pinging an entity update does the trick.

This does not work with the alarm/timers etc tho, they are all still on unknown

**Comment by jlong23** [comment]
> > Does running homeassistant.update_entity on the relevant media_player entity help at all?
> 
> This did it for me, had the same problem, but manually pinging an entity update does the trick.

This seems to help, kept updating the entity from that point while playing, stopped playback and started a new Pandora Session and it still updated.   Will play if we need to schedule the service call for periodic refreshing of the 7+ AMZ Devices

**Comment by krisnoble** [comment]
> > > Does running homeassistant.update_entity on the relevant media_player entity help at all?
> > 
> > 
> > This did it for me, had the same problem, but manually pinging an entity update does the trick.
> 
> This does not work with the alarm/timers etc tho, they are all still on unknown

Just checking, you're running the update service on the relevant sensor entity rather than the media_player? That works for me although could be one of those tempramental ones where sometimes it works, sometimes it doesn't.

You can run the following in the developer tools template section to get a YAML-formatted list of all enabled entities from the integration - I can't figure out how to get it to work in the actual service call right now but you can copy-paste and save some typing at least...

```
{% for entity in integration_entities("alexa_media") %}
- {{ entity }}
{%- endfor %}
```

**Comment by krisnoble** [comment]
> > > Does running homeassistant.update_entity on the relevant media_player entity help at all?
> > 
> > 
> > This did it for me, had the same problem, but manually pinging an entity update does the trick.
> 
> This seems to help, kept updating the entity from that point while playing, stopped playback and started a new Pandora Session and it still updated. Will play if we need to schedule the service call for periodic refreshing of the 7+ AMZ Devices

Yeah, I have the following running, not sure if overkill but seems to run okay and keeps things updated:

```
alias: Alexa media updates
description: ""
trigger:
  - platform: homeassistant
    event: start
condition: []
action:
  - repeat:
      until:
        - condition: state
          entity_id: input_boolean.ha_restarting
          state: "on"
      sequence:
        - service: homeassistant.update_entity
          data: {}
          target:
            entity_id:
              - < list of entities >
        - delay:
            hours: 0
            minutes: 0
            seconds: 10
            milliseconds: 0
mode: restart
```

**Comment by jlong23** [comment]
I left the dot idle for a few hours and started a new Pandora session, didn't update until I called the
`service: homeassistant.update_entity
data: {}
target:
  entity_id: media_player.computer_room_dot
`

Suspect that it falls off the polling when idle for a while.

Will implement the suggested trigger that @krisnoble posted, probably set it for repeat at 1 min as my use-case is watching / recording the current song played on Pandora when I am in the shower.

**Comment by emufan** [comment]
> but manually pinging an entity update does the trick.

But where is the trick, if I don't know, when a timer is running. You are suggesting, to poll/ping every entity periodically e.g. every 1 minute 24/7? I don't know, if this is a good apprach.

**Comment by jlong23** [comment]
> You are suggesting, to poll/ping every entity periodically e.g. every 1 minute 24/7? I don't know, if this is a good apprach.

Agreed, but a production / user workaround is better than nothing.
I would like a bit more control over the the HA Polling threads, or Ubiquity UDME not to loose fixed DNS entries when it shouldn't and forget ones when it should, no defects there  ;) 

The polling the monitored media_players is working for me ATM, 12% Utilization on the NUC style HA host (No noticeable Changes).  So I am happy now that I can see my last ten items on my playlist dashboard.

**Comment by krisnoble** [comment]
> You are suggesting, to poll/ping every entity periodically e.g. every 1 minute 24/7? I don't know, if this is a good apprach.

FWIW this is my system running in Docker on a Pi4 4GB, updating every 10 seconds. The raised periods are in the evenings when the Pi runs some other stuff like a slideshow and some custom scripts. Peak values are 58c, 19% processor, 56% memory, 1.4 load average, all well within nominal as far as I know. Hopefully that's reassuring?

<img width="1006" alt="Screenshot 2023-07-26 at 13 41 22" src="https://github.com/custom-components/alexa_media_player/assets/872971/7c210bdc-636e-4075-a27a-3a9d539ccd5b">

**Comment by alandtse** [comment]
FYI, the concern with excessive polling is that Amazon will throttle or block you. This is the reason we do not use that as a solution generally. But if you're ok with that risk, this may be a workaround.

**Comment by DarkStarSword** [comment]
Agree with alandtse on that, however it should be noted that the risk could potentially extend beyond those who are using the workaround to all users of Alexa Media Player depending on what Amazon chooses to do about it if/when they notice.

Presumably the reason this broke in the first place was because Amazon was trying to reduce their API load and/or traffic and disabled broadcasting state changes to achieve that (or it could just be a bug and completely unintentional, but we have no way to know that). Now that some users are polling them for state that would increase their traffic rather than reducing it, and once they notice that and their investigation points to Alexa Media Player, they may either block the specific users, or may just block the entire app to stop the increased load.

The thing is, if they don't pay very close attention they may not even realise that it is only a few users causing it which means the chance that they block the entire app is higher than you might naively expect - however it is hard to say how big that risk actually is.

**Comment by krisnoble** [comment]
Good points both and apologies, I didn't realise the update polling wasn't purely local. I certainly don't want to cause any problems for anyone so I've dropped my polling down to once an hour for now which has been stable all morning as far as I know and should be much more reasonable in Amazon's eyes. 

I was already planning to migrate from Alexa to something more open anyway, so that plan has just been shifted from "soon" to "now". FWIW though thank you Alan et al for all your work on this, it's been immensely helpful over the last few years and much appreciated.

**Comment by Pirol62** [comment]
Hi, 
facing the same issue I'd like to refresh only the selected alexa group when music playing is active once a minute.
Would that be acceptable?

````
- id: '206'
  alias: Alexa media player updates
  description: ""
  trigger:
    - platform: state
      entity_id: input_boolean.music_playing
      to: 'on'
  condition: []
  action:
    - repeat:
        while:
          - condition: template
            value_template: "{{ is_state('input_boolean.music_playing','on') }}"
        sequence:
          - service: homeassistant.update_entity
            data: {}
            target:
              entity_id: >
                {% set player = states('input_select.music_sources') %}
                {% set map = {
                  "Erdgeschoss": "media_player.erdgeschoss",
                  "Alle Etagen": "media_player.alle_etagen",
                  "Home Office": "media_player.home_office",
                  "Dinnerparty": "media_player.dinnerparty"
                }%}
                {{ map[player] if player in map else 'media_player.receiver' }}
          - delay:
              hours: 0
              minutes: 1
              seconds: 0
              milliseconds: 0
  mode: restart
````

**Comment by krisnoble** [comment]
> Hi, facing the same issue I'd like to refresh only the selected alexa group when music playing is active once a minute. 

The trouble with that approach (which is what I first thought of too) is that if the device does get into the "stuck" state (e.g. been left for too long as posited by jlong23), it will never get the "playing" state update in HA to trigger the automation. Looking in the code it seems like the default polling interval is one minute anyway (and this tallies with my own observations) so as long as the devices stay active I don't think manually polling once per minute would bring any benefit.

**Comment by Pirol62** [comment]
PS: I tested it over the weekend with 5 Minutes delay and it worked stable. Interestingly the media player obvously stops updating after 3 or 4 Minutes and only sometimes. So I will set the delay to 3 Minutes. That seem to be good in order to have an accurate display of the current state.

**Comment by DarkStarSword** [comment]
@Pirol62 While I can't speak for what thresholds Amazon deems acceptable, that looks like an entirely reasonable workaround to me to keep polling to a minimum while still allowing the desired functionality.

**Comment by energywave** [comment]
I also have the same problem and I'm constantly being contacted from people that are ousing my [multinotify](https://github.com/energywave/multinotify) that are having this issue that generates two issues:

- the volume is not resumed (as it stopped updating with the real value)
- the music play status is wrong so at the end of the message echo will not resume the play or will play while it was not playing before (as the status was not updated), quite annoying.

Generally the "solution" is to reload the integration or restart Home Assistant but not having even a way to detect that this is happening is very bad.

From my perspective it could be that the websocket is closing without notice so it stops to receive updates from Amazon.
I'm not a python programmer so I'm not able to contribute (sigh...) but couldn't anyone check the webserver management code? We could work on that to ensure the websocket is alive and if it's not we can reestablish it.
As a last resort a code that every time interval recycle the websocket could be a temporary fix that helps this not to happen too frequently. But websocket closing should be a very detectable event that should be handled.

**Comment by PaulBiod** [comment]
Hi,
I am experiencing the same bug, for some reasons, the media player state isn't updating at all, and only a reload of the integration can fix the bug, I don't know for how much time. This is really problematic because many automations depend on these states.
I'm not a python coder but I wish someone will put his hands on it to fix it, cause this integration is so popular.
Anyway, thanks to  the author, and I hope this bug can be fixed in the future.

**Comment by steff75** [comment]
I'm not sure if it's the same or at least a similar issue, but competitor addons have also had issues recently (ioBroker, node-red) after changes at Amazon. These are both dependent on the [alexa-remote2 library](https://github.com/Apollon77/alexa-remote), where the developer has now made some adjustments to make it work again temporarily.

**Comment by energywave** [comment]
> .

Hello if that issue started recently I don't think it's the same as this issue is present from ever, years at least.

**Comment by DarkStarSword** [comment]
> this issue is present from ever, years at least.

You sure about that? This ticket was only opened in the last few months:

![image](https://github.com/custom-components/alexa_media_player/assets/379749/966b83fc-fa6f-4b32-a147-bff515075d95)

I also only started encountering this issue around that time, give or take. If the issue existed before that perhaps something has made it get worse recently so more people started encountering it / noticing?

> From my perspective it could be that the websocket is closing without notice so it stops to receive updates from Amazon.

That's not how I interpreted Alan's analysis of the log file, unless I misunderstood?

https://github.com/custom-components/alexa_media_player/issues/1953#issuecomment-1579807928

**Comment by emufan** [comment]
Here it starts only in May as well. But I have the feeling that it is getting even worse in the last days and week. My timers are now not updating anymore in 90%. Not when they are started and (this is new) not even when they are then finished/outdated.

**Comment by energywave** [comment]
> > this issue is present from ever, years at least.
> 
> You sure about that? This ticket was only opened in the last few months

Yes, I'm absolutely sure about that. There was previous issues about this argument that was closed, here the first I've found where I've participating in: https://github.com/custom-components/alexa_media_player/issues/1578
That's more than a year ago (and the problem was there from well before).

> I also only started encountering this issue around that time, give or take. If the issue existed before that perhaps something has made it get worse recently so more people started encountering it / noticing?

I don't know why that problem arrives in waves, maybe when Amazon changes something then it happens more frequently, who knows.
In fact if you don't use volume_level or timers, or play status of Alexa devices there a good chance you'll even not noticing this issue is there as it's not preventing announcements and other commands from being run. It makes status updates from Amazon not arriving so, for example, you change the volume of a device by voice but that volume in the Home Assistant entity don't change.
I noticed that from much time as I've developed [multinotify](https://github.com/energywave/multinotify), a script that, between the other functions, backup the volume and play status of devices, play the announcement then restore the volume of every device and resume the play where it was paused.
When the websocket stops working that script is backing up a not real status and that makes you understand to be in that condition easily.

> 
> > From my perspective it could be that the websocket is closing without notice so it stops to receive updates from Amazon.
> 
> That's not how I interpreted Alan's analysis of the log file, unless I misunderstood?
> 
> [#1953 (comment)](https://github.com/custom-components/alexa_media_player/issues/1953#issuecomment-1579807928)

Yes, you're right. This could be a brand new issue with same symptoms that was there from ever. But stating that my Alexa devices are working as before (they update with the Echo devices status then, after a random amount of time, they stops updating) I thought that this issue was related to that condition that there from ages.

**Comment by UglyBob79** [comment]
Similar issue here, I got my Alexa devices a few weeks ago and I _NEVER_ got the players to update state in Home assistant. I play music by telling Alexa to play and all media_player entities are still in standby. I have reinstalled the integration multiple times, nothing helps. I have an Echo Show 5, an Echo Studio and a Wiim Pro, none update status.

**Comment by Pirol62** [comment]
Today I detected the first time, that timers are not updated anymore. Restart HA or refresh entities has no effect.

**Comment by emufan** [comment]
Here `homeassistant.update_entity` is working sometimes. I seems that it is only working some minutes after starting a timer and some minutes or hours.

**Comment by Pirol62** [comment]
Ok, obviously the timers are not supported in the interface by amazon any more. So I deactivated all entities and took them out of the views. As I had no further automation on these entities, it is ok for me.  It was sometimes nice to see, that a timer is active but so what...
The issue with the media players is more relevant but works for my with my workaround see 14 posts above.
In addition to that I relaod alexa once a night. That clears the entities from old states which are no longer correct.

I everything stayed like this..I can live with it.
I hope that amazon will not reduce further functions because reminders and alarms have a relevance for me.

**Comment by LePremier** [comment]
Same issue here. Restarting the service once a day (it is part of a morning routine I run atomically) helped for some time, but now also this is not anymore the case :(

**Comment by Ulli2k** [comment]
> Ok, obviously the timers are not supported in the interface by amazon any more. So I deactivated all entities and took them out of the views. As I had no further automation on these entities, it is ok for me. It was sometimes nice to see, that a timer is active but so what... The issue with the media players is more relevant but works for my with my workaround see 14 posts above. In addition to that I relaod alexa once a night. That clears the entities from old states which are no longer correct.
> 
> If everything stayed like this..I can live with it. I hope that amazon will not reduce further functions because reminders and alarms have a relevance for me.

I solved my personal issue of alarms with restarting the instance periodically.
Therefore removing the entitys would be not the best solution for some of us.

I understood amazon skipped the event based functionality of reporting entity values. But by restarting the integration it is still possible to read the current alarm values.
=> why not implementing a poll function  which peroidically requests the values. Or make a service availabe for the useres to force updating the values manually?

**Comment by HeroCity99** [comment]
Today I noticed that I am having the same issue where the media player (Alexa) status isn't updating for any of its attributes unless the entity is reloaded manually. Ive created an automation to overcome this by making the trigger for any status change in the media player to reload the "Home Assistant Core Integration: Update entity". This works but isn't really a solution.

**Comment by alandtse** [comment]
> => why not implementing a poll function which peroidically requests the values. Or make a service availabe for the useres to force updating the values manually?

https://github.com/custom-components/alexa_media_player/issues/1953#issuecomment-1652111938

I wish there was a way to sticky a post.

**Comment by ollisulopuisto** [comment]
Just to add a bit of context: 

My HA setup shows the same issue, so I implemented a periodical update_entity–thanks for the idea!

Looking at the logs, it would appear that media_position is updated every 20 seconds or so as long as a track is playing, but when the track ends and a new one starts, the updating stops. (I was thinking of creating an automation that would trigger update_entity when media_duration equals media_position, but alas no such luck.)

I guess you could somehow calculate the right moment to trigger an update by comparing media_duration, media_position and media_position_updated_at?

**Comment by emufan** [comment]
I already tried to use update_entity EG after starting a timer or after the timer is finished. But without success. The update_entity ist not reliable and not pulling the right state most of the time. Even without error it is only delivering the right state sometimes. In the example e.g. only random 5-15 minutes after starting or ending a timer. Why ever.

**Comment by Ulli2k** [comment]
> I already tried to use update_entity EG after starting a timer or after the timer is finished. But without success. The update_entity ist not reliable and not pulling the right state most of the time. Even without error it is only delivering the right state sometimes. In the example e.g. only random 5-15 minutes after starting or ending a timer. Why ever.

Did you use > homeassistant.reload_config_entry

**Comment by formatBCE** [comment]
Same here. Will wait for updates. If any help is needed, please reach out.

**Comment by darkprojekt** [comment]
Seems that this integration is going to be abandoned very soon, when such fundamental issues are not fixed anymore. I really loved it.

**Comment by broyuken** [comment]
> Seems that this integration is going to be abandoned very soon, when such fundamental issues are not fixed anymore. I really loved it.

I agree, I’m hoping that nabu casa comes out with some hardware at the end of the year of the voice to compete with Alexa, then we can let this integration die

**Comment by M3d1c5** [comment]
> I agree, I’m hoping that nabu casa comes out with some hardware at the end of the year of the voice to compete with Alexa, then we can let this integration die

You seem to have misunderstood something here.

This custom component is not used to transmit voice commands to Home Assistant.
This custom component is used to control Amazon Alexa and Amazon Echo devices.

Voice commands are transmitted to Home Assistant via the Nabu Casa Skill and the Nabu Casa Cloud.

**Comment by broyuken** [comment]
I don’t misunderstand anything. This is the year of the voice, and my hope is that they come up with hardware that communicates directly with THEIR assist assistant so I can get rid of my Alexa devices and their reliance on Amazon and their cloud services to work.

**Comment by 1liminal1** [comment]
> This is the year of the voice, and my hope is that they come up with hardware that communicates directly

I dont think thats going to happen anytime soon :(

**Comment by broyuken** [comment]
> > This is the year of the voice, and my hope is that they come up with hardware that communicates directly
> 
> I dont think thats going to happen anytime soon :(

Why not? They already have voice integrated with a number of other devices. All they need to implement is a wake word and we’re in business. The Muse LUXE speaker already exists, running esphome firmware. I think it’s very reasonable to expect something like this.

**Comment by 1liminal1** [comment]
Eventually yes. But the wake word is the super hard part apparently

**Comment by herveaurel** [comment]
but for now, still having problems with Alexa Media Player and entities that are not updating... What a shame to no longer see the music, alarms and other things!

**Comment by 1liminal1** [comment]
This might help someone that uses Spotify and Node Red I used the swtich node to get the source 

![image](https://github.com/custom-components/alexa_media_player/assets/10505942/5d657102-bd7c-4a09-8f29-b6c9dc563839)

**Comment by formatBCE** [comment]
> This might help someone that uses Spotify and Node Red I used the swtich node to get the source 
> 
> ![image](https://github.com/custom-components/alexa_media_player/assets/10505942/5d657102-bd7c-4a09-8f29-b6c9dc563839)
> 

1. This issue is about overall not updating state. How using attribute could help?
2. Why use NodeRED for simple state attribute, if you can create template sensor, or use it directly in automation?

Either you're bot/troll, or you got mistaken with the issue.

**Comment by 1liminal1** [comment]
LOL ok... or it depends on how you're using the states. I use it for follow me music. I'm saying, that if you're using it to get the media state on a specific Spotify device. You can use a switch node to pass that info along. 

Its not for every use case, I said that :) 

And why would you go to the trouble of creating a template sensor for something as simple as passing a state attribute ;)  

But thank you for letting me know that it didn't apply to you, I appreciate it big guy!

**Comment by formatBCE** [comment]
> LOL ok... or it depends on how you're using the states. I use it for follow me music. I'm saying, that if you're using it to get the media state on a specific Spotify device. You can use a switch node to pass that info along. 
> 
> Its not for every use case, I said that :) 
> 
> And why would you go to the trouble of creating a template sensor for something as simple as passing a state attribute ;)  
> 
> 

Still don't need NodeRED for it. I'm saying because I used to use RED extensively in the past.

For our problem: state doesn't have an update AT ALL, so no changes to attributes as well...

**Comment by steff75** [comment]
But you can still use Node-RED here as a (temporary) workaround to get the events.

This is done with the node-red-contrib-alexa-remote2-applestrudel node, which gets its functionality from the alexa-remote2 lib.

And this alexa-remote2 lib is mostly updated in time by its developer when changes from Amazon cause problems. Like the ones from the last few weeks. You only have to activate the events in the settings of the node, because this is deactivated by default.

As a bridge between Node-RED and Home Assistant you can use MQTT or the node-red-contrib-home-assistant-websocket node.

**Comment by formatBCE** [comment]
Got you. So NodeRED lib works at least.

Not that I want to do it, as it will require to re-do everything related to Alexa.. but I see your point!

**Comment by 1liminal1** [comment]
What I'm saying is its clear this integration is dead. The dev does not want to fix it, so we all need to figure out other ways to make get the state values we need. 

I know its a hassle, I've brought this guy a "coffee" before. But that's just the way it goes. I got a couple of great years out of it, so it was worth it. Now we have to figure out something else. If I was a dev, believe me I would have been fixing this already LOL.

**Comment by formatBCE** [comment]
Let's ask @alandtse if this is something that's on roadmap. Last release was 3 days ago - so maybe it's not lost yet.

**Comment by alandtse** [comment]
Given the trends, I'm expecting for Amazon to shut down the ecosystem soon. So not in my roadmap but I will take PRs of people figure it out without just spiking the polling. That's why it's tagged as help wanted.

**Comment by formatBCE** [comment]
@alandtse thanks for quick reply. I understand, that Amazon can shut that door once and forever (given it's still open, of course)...
It's been a while since I was tinkering with Python, but I'll try to dig it a bit, if you put my face in most low-level API file, please. :)

**Comment by 1liminal1** [comment]
Can I just ask as well, when you say ecosystem. Are you talking about the media player portion of that? So Alexa will stop allowing people to play music. Or the whole thing?

**Comment by formatBCE** [comment]
This integration is using unofficial API. It's nothing to do with Alexa functionality - it's about third-party access to Alexa functions. And it's not guaranteed to work in any way.

**Comment by 1liminal1** [comment]
Ahh ok got ya. thanks :)

**Comment by Profex** [comment]
Openhab has also currently issues with their integration (Last Voice command no longer reported), and one of the maintainers said its due to Amazon is no longer reporting last voice command state changes via websocket. Instead Amazon is now reporting via HTTP/2 stream. Could it be that this linked, and the same is true for the media player states?

https://community.openhab.org/t/amazon-echo-smartj-binding-last-voice-command-not-working-anymore/148296/68

**Comment by cpadil** [comment]
Seems you are right, in fact the "alexa-remote" project (javascript based api) switched to HTTP2 streams a few weeks ago
https://github.com/Apollon77/alexa-remote/releases/tag/v6.0.0

an interesting fact is that this project "alexa media player" is using the python package "alexapy" (https://gitlab.com/keatontaylor/alexapy) which is based on the "alexa-remote" mentioned above

there are some python projects implementing https2 streams, e.g. https://github.com/python-hyper/h2 however 

however seems a complex change

**Comment by broyuken** [comment]
Hopefully someone smarter than I can put in a PR to bring back this functionality. This sounds likely that this is the cause of the issues we're all experiencing. If there's anything I can do to assist I will gladly, but python is a little outside of my ability.

**Comment by formatBCE** [comment]
I tried proposed NodeRED alternative - and it's actually lightning-fast in TTS announcements, skills like Actionable Notifications etc. No subscriptions though, as far as I understand. So returning to status-quo would be preferable - and then I can slowly move some functionality, not related to media player itself, to NodeRED. :)

**Comment by 1liminal1** [comment]
Ive never not used Node Red, got into it early on and never looked back lol. 

I dont think I gave enough info on how Im using it. 

This is one example. So now that Alexa devices no longer reliably report current state as checks, I'm now using Spotify entity and then the switch node to filter for Alexa devices. 

![image](https://github.com/custom-components/alexa_media_player/assets/10505942/2cd45b44-0a91-47c5-ad3e-5c9e704dc545)

But the TTS is also great, I use it to let me know when prints are finished, % of prints, that particular windows need to be closed before turning on the aircon, washing machines finished, I could go on lol

**Comment by formatBCE** [comment]
I mean push state updates. Exactly what we've lost here  :)

**Comment by alandtse** [comment]
https://github.com/Apollon77/alexa-remote/compare/v5.10.3...v6.0.0 shows the changes to enable http2 pushes. Unfortunately, alexapy is built on [aiohttp, which does not support http2](https://docs.aiohttp.org/en/stable/changes.html#id570). 

[Httpx does support http2](https://www.python-httpx.org/http2/). However, swapping from aiohttp to httpx is a large undertaking. Perhaps it can be done by just replacing the websocket with httpx. If someone submits a PR to [gitlab](https://gitlab.com/keatontaylor/alexapy/) to address this, ping me here. I don't check gitlab anymore.

**Comment by oneseventhree** [comment]
> I tried proposed NodeRED alternative - and it's actually lightning-fast in TTS announcements, skills like Actionable Notifications etc. No subscriptions though, as far as I understand. So returning to status-quo would be preferable - and then I can slowly move some functionality, not related to media player itself, to NodeRED. :)

Does Node Red show last_called and timers?

**Comment by formatBCE** [comment]
> > I tried proposed NodeRED alternative - and it's actually lightning-fast in TTS announcements, skills like Actionable Notifications etc. No subscriptions though, as far as I understand. So returning to status-quo would be preferable - and then I can slowly move some functionality, not related to media player itself, to NodeRED. :)
> 
> Does Node Red show last_called and timers?

Last called can be retrieved by getting Activities and picking the latest one. It does even tell what was the request.

Timers/alarms are returned with Notifications for device.

You can play with it yourself, it's easy.

**Comment by oneseventhree** [comment]
> > > I tried proposed NodeRED alternative - and it's actually lightning-fast in TTS announcements, skills like Actionable Notifications etc. No subscriptions though, as far as I understand. So returning to status-quo would be preferable - and then I can slowly move some functionality, not related to media player itself, to NodeRED. :)
> > 
> > 
> > Does Node Red show last_called and timers?
> 
> Last called can be retrieved by getting Activities and picking the latest one. It does even tell what was the request.
> 
> Timers/alarms are returned with Notifications for device.
> 
> You can play with it yourself, it's easy.

Any YouTube/other guides you recommend? I’ve never played around with node red before

**Comment by formatBCE** [comment]
> > > > I tried proposed NodeRED alternative - and it's actually lightning-fast in TTS announcements, skills like Actionable Notifications etc. No subscriptions though, as far as I understand. So returning to status-quo would be preferable - and then I can slowly move some functionality, not related to media player itself, to NodeRED. :)
> > > 
> > > 
> > > Does Node Red show last_called and timers?
> > 
> > Last called can be retrieved by getting Activities and picking the latest one. It does even tell what was the request.
> > 
> > Timers/alarms are returned with Notifications for device.
> > 
> > You can play with it yourself, it's easy.
> 
> Any YouTube/other guides you recommend? I’ve never played around with node red before

Here you go, one of my favourite YouTubers: https://youtu.be/hBEb_FCLRU8?si=7YoE6HbN9ExgtijF

**Comment by tebbbb** [comment]
This fixes this problem for me. It reloads all my Alexas every 30 seconds. Replace my alexa entities with your own and you are good to go.


alias: Home | Reload all Alexas
description: ""
trigger:
  - platform: time_pattern
    hours: "*"
    minutes: "*"
    seconds: /30
condition: []
action:
  - service: homeassistant.reload_config_entry
    data: {}
    target:
      entity_id:
        - media_player.lounge
        - media_player.echo_isabella
        - media_player.echo_dot_bathroom
        - media_player.echo_dot_bedroom
        - media_player.echo_show
        - media_player.echo_living_room
mode: single

**Comment by formatBCE** [comment]
> This fixes this problem for me. It reloads all my Alexas every 30 seconds. Replace my alexa entities with your own and you are good to go.
> 
> ```
> alias: Home | Reload all Alexas
> description: ""
> trigger:
>   - platform: time_pattern
>     hours: "*"
>     minutes: "*"
>     seconds: /30
> condition: []
> action:
>   - service: homeassistant.reload_config_entry
>     data: {}
>     target:
>       entity_id:
>         - media_player.lounge
>         - media_player.echo_isabella
>         - media_player.echo_dot_bathroom
>         - media_player.echo_dot_bedroom
>         - media_player.echo_show
>         - media_player.echo_living_room
> mode: single
> ```

Great it works for you. However, it's partial, and really bug-prone solution. Also, you don't need to put all entities there - they all are under one config entry, so putting one media player I'd would be enough to reload all of them.

**Comment by Kethlak** [comment]
@formatBCE New to Node Red here, installed and configured based on your video recommendation (thanks!). Is there a way to set up a trigger for an Alexa timer being set? I looked at the device trigger node, but it doesn't seem to be what I need.

**Comment by formatBCE** [comment]
> @formatBCE New to Node Red here, installed and configured based on your video recommendation (thanks!). Is there a way to set up a trigger for an Alexa timer being set? I looked at the device trigger node, but it doesn't seem to be what I need.

That's what issue currently is about. 
You see, there's two types of dealing with information updates: pushing and polling. Pushing is great, and allows information to update as it goes (like, timer was set -> you received push trigger on it). But it requires Alexa to have a mechanism for subscription to that updates. That's what has changed lately - and that's why our integration doesn't work.
In case of NodeRED and Alexa integration - I don't see any signs of push subscription possibility. So you'd have to use polling.

Polling is easier to do, but it has huge disadvantages:
1. It doesn't update instantly, but only on next poll cycle. So if you're polling each 10 seconds, you can completely miss 9-second timer, for example.
2. If you make polling period shorter, it will greatly load Alexa API  (and, probably, your server), and most of that requests will be wasting of computing capacity, because no new information will be polled...

Long story short: you can poll. No push. I won't recommend that. :)

**Comment by Kethlak** [comment]
@formatBCE Ah, thanks. I was hoping NodeRED would let me get around the issue, but not in my case. Bummer.

I wish it was easier to set up an environment to write addons for HA, then I might be able to look into the aiohttp to httpx conversion and help out with the plugin. I'm primarily a web developer and I'm used to working with C#, PHP, and JavaScript, but I have a tiny bit of Python experience. I keep thinking of helping out, but the barrier to entry seems pretty high.

**Comment by formatBCE** [comment]
@Kethlak well, you don't need to write add-on in this case - it's just custom integration, so basically you can test it by changing the code directly inside of custom_components folder in Home Assistant. :)

The hardest part, I guess, is tinkering with your own Amazon account...

**Comment by broyuken** [comment]
> > This fixes this problem for me. It reloads all my Alexas every 30 seconds. Replace my alexa entities with your own and you are good to go.
> > ```
> > alias: Home | Reload all Alexas
> > description: ""
> > trigger:
> >   - platform: time_pattern
> >     hours: "*"
> >     minutes: "*"
> >     seconds: /30
> > condition: []
> > action:
> >   - service: homeassistant.reload_config_entry
> >     data: {}
> >     target:
> >       entity_id:
> >         - media_player.lounge
> >         - media_player.echo_isabella
> >         - media_player.echo_dot_bathroom
> >         - media_player.echo_dot_bedroom
> >         - media_player.echo_show
> >         - media_player.echo_living_room
> > mode: single
> > ```
> 
> Great it works for you. However, it's partial, and really bug-prone solution. Also, you don't need to put all entities there - they all are under one config entry, so putting one media player I'd would be enough to reload all of them.
> 
> P.S. also, while it's reloading, devices are unavailable - and some automations, that use these devices, can be broken. I'm assuming you're not using Alexa for important TTS notifications.

Works great for me and it doesn’t make them go unavailable. We’re just updating the entities, not reloading the integration. This seems to be a decent workaround for now.

**Comment by emufan** [comment]
> Works great for me and it doesn’t make them go unavailable. We’re just updating the entities, not reloading the integration. This seems to be a decent workaround for now.

You are still wrong here. reload_config_entry is reloading the integration. See https://www.home-assistant.io/integrations/homeassistant/#service-homeassistantreload_config_entry.

Updating an entities is via update_entity. See https://www.home-assistant.io/integrations/homeassistant/#service-homeassistantreload_config_entry.

**Comment by broyuken** [comment]
My fault, my automation is using update entity.

**Comment by popy2k14** [comment]
> My fault, my automation is using update entity.

Can you please provide your automation which doesn't restart the integration but can reload/update the entity values?

**Comment by amaisano** [comment]
Here is a script that will update an entity, but it rarely works for me. The timers continue to be Unknown no matter how many times I spam it:

```yml
alias: Reload Alexa Library Timer
sequence:
  - service: homeassistant.update_entity
    data: {}
    target:
      entity_id: sensor.library_dot_next_timer
mode: restart
```

Perhaps I'm targeting the wrong kind of entity for this workaround?

**Comment by broyuken** [comment]
alias: "☑️ Misc: Update Alexa Devices"
description: ""
trigger:
  - platform: time_pattern
    seconds: /30
condition: []
action:
  - service: homeassistant.update_entity
    data: {}
    target:
      entity_id:
        - media_player.office_echo
        - media_player.gym_echo_dot
        - media_player.entryway_echo_dot
        - media_player.playroom_echo_dot
        - media_player.downstairs_bathroom_echo_dot
        - media_player.kitchen_echo_show
        - media_player.living_room_echo
        - media_player.master_bathroom_echo_dot
        - media_player.kids_bathroom_echo_dot
        - media_player.justins_echo_dot
        - media_player.charlottes_echo_spot
        - sensor.master_bedroom_echo_show_next_alarm
        - media_player.master_bedroom_echo_show
        - sensor.charlottes_echo_spot_next_alarm
mode: single

**Comment by emufan** [comment]
From my experiences, every 30 seconds is not needed at all, because the service only delivers new status ~ every 10 minutes +/-. At least for the timers. Whyever.

**Comment by popy2k14** [comment]
> alias: "☑️ Misc: Update Alexa Devices" description: "" trigger:
> 
> * platform: time_pattern
>   seconds: /30
>   condition: []
>   action:
> * service: homeassistant.update_entity
>   data: {}
>   target:
>   entity_id:
>   - media_player.office_echo
>   - media_player.gym_echo_dot
>   - media_player.entryway_echo_dot
>   - media_player.playroom_echo_dot
>   - media_player.downstairs_bathroom_echo_dot
>   - media_player.kitchen_echo_show
>   - media_player.living_room_echo
>   - media_player.master_bathroom_echo_dot
>   - media_player.kids_bathroom_echo_dot
>   - media_player.justins_echo_dot
>   - media_player.charlottes_echo_spot
>   - sensor.master_bedroom_echo_show_next_alarm
>   - media_player.master_bedroom_echo_show
>   - sensor.charlottes_echo_spot_next_alarm
>   mode: single

thx a lot, that did it!!!!!! FInally i can use the integration again!

> From my experiences, every 30 seconds is not needed at all, because the service only delivers new status ~ every 10 minutes +/-. At least for the timers. Whyever.

Not true when you example change the volume of the echo.

**Comment by popy2k14** [comment]
Spoke too soon ... updates worked instantly for a device the first times. But now now updating again :-(

**Comment by formatBCE** [comment]
> Spoke too soon ... updates worked instantly for a device the first times. But now now updating again :-(

Because you're bombing Alexa API with requests each 30 seconds, and I guess after some time API starts to debounce it. That's why polling is bad, and these rude solutions are worse than no solution.

**Comment by WhimsySpoon** [comment]
> polling is bad, and these rude solutions are worse than no solution

This. 100%.

Unfortunately it seems the only real solution here is to he the new HTTP/2 API.

**Comment by alandtse** [comment]
Ok, I'm trying to put actual information in this thread to try to resolve this. However, with all the chatter, it's going to get buried. Please take the workaround chatter to the [discussion forum](https://github.com/custom-components/alexa_media_player/discussions/2037). 

Also, [can I get a volunteer to go through this thread and help clean it up and maybe more moderation](https://github.com/custom-components/alexa_media_player/discussions/2038)? 

Anyway back to the technical:

- The AWS http2 protocol may actually be https://developer.amazon.com/en-US/docs/alexa/alexa-voice-service/http2-overview.html.
- At least one person tried to get httpx's http2 to work with AWS. They deleted their repo though but there may be more pointers through this issue and the linked issues. https://github.com/encode/httpx/issues/1410

**Comment by alandtse** [comment]
This may be useful. https://github.com/mkb79/Audible. Apparently for audible registration `access_token` got deprecated.

Regardless, I may be able to do a http2 release but users will have to manually generate an `access_token`. https://github.com/Apollon77/alexa-remote. For some reason the way we generate tokens does not allow http2 connections. It may very well be my account is blocked due to too many connections.

**Comment by WhimsySpoon** [comment]
Is there anything we can do to help? Would it be beneficial if we tried to generate a token to confirm your blocked account theory?

**Comment by starkillerOG** [comment]
@alandtse am I reading correctly that [alexa-remote](https://github.com/Apollon77/alexa-remote) is using a local connection to the alexa devices instead of a cloud connection such as this library is currently using (I think)?

**Comment by rumunn95** [comment]
> Ok, I'm trying to put actual information in this thread to try to resolve this. However, with all the chatter, it's going to get buried. Please take the workaround chatter to the [discussion forum](https://github.com/custom-components/alexa_media_player/discussions/2037).
> 
> Also, [can I get a volunteer to go through this thread and help clean it up and maybe more moderation](https://github.com/custom-components/alexa_media_player/discussions/2038)?
> 
> Anyway back to the technical:
> 
> * The AWS http2 protocol may actually be https://developer.amazon.com/en-US/docs/alexa/alexa-voice-service/http2-overview.html.
> * At least one person tried to get httpx's http2 to work with AWS. They deleted their repo though but there may be more pointers through this issue and the linked issues. [HTTP2 Sync Client blocked after attempting to establish a downchannel to Alexa Voice Service encode/httpx#1410](https://github.com/encode/httpx/issues/1410)

look to discussion i paste my config from node red now playing status is refreshing

**Comment by WhimsySpoon** [comment]
Thank you for getting this updated so quickly!

I've installed the latest version, restarted HA and had an Echo Dot play music for +5 minutes. The media_player entity didn't update during the time and this was the only entry in the log:

````
2023-09-29 09:48:41.068 DEBUG (MainThread) [alexapy.alexahttp2] Preparing ping to https://bob-dispatch-prod-na.amazon.com/ping
2023-09-29 09:48:41.159 DEBUG (MainThread) [alexapy.alexahttp2] Received response: 204:
````

Is there something else that I should be doing?

**Comment by Profex** [comment]
Thanks as well from my side!

But can not confirm what WhimsySpoon mentioned .. just updated and tested as well, and for me it's working perfectly fine again!

**Comment by alandtse** [comment]
It's possible you have to relogin to force a new access token to get generated. Remove and readd the integration,.

**Comment by francescolf** [comment]
Even after relogging, push updates don't work. This might be related:

```
2023-09-29 11:29:48.092 DEBUG (MainThread) [alexapy.alexahttp2] Received raw message: --------abcde123
2023-09-29 11:29:48.112 DEBUG (MainThread) [alexapy.alexahttp2] HTTP2 Connection Closed.
2023-09-29 11:29:48.113 DEBUG (MainThread) [alexapy.alexahttp2] HTTP2 Connection Closed.
2023-09-29 11:29:48.114 DEBUG (MainThread) [custom_components.alexa_media] c*********@g*******m: Close requested; will not reconnect http2
2023-09-29 11:29:48.115 DEBUG (MainThread) [custom_components.alexa_media] c*********@g*******m: Close requested; will not reconnect http2
```

**Comment by popy2k14** [comment]
Also for me after removing the integration and readd/relogin, media players does not update.
I have enbled debug logging for the compontent, but there was no usefull information why it fails.
The initial fetch of data works.

Anything we can provide to help out?

**Comment by WhimsySpoon** [comment]
I don't remember seeing this Public URL option before. Is it relevant? I've populated it but it hasn't changed anything, but I won't be in a position to remove and re-add the integration until later today.

![image](https://github.com/custom-components/alexa_media_player/assets/19993138/2a7a472c-045e-487c-a9bc-d172baae116f)

**Comment by oneseventhree** [comment]
> Also for me after removing the integration and readd/relogin, media players does not update. I have enbled debug logging for the compontent, but there was no usefull information why it fails. The initial fetch of data works.
> 
> Anything we can provide to help out?

same as me. nothing in logs but all entites are unavailable:
<img width="319" alt="image" src="https://github.com/custom-components/alexa_media_player/assets/15043282/e0b8ff66-1f97-4d4d-91a3-8cd68c61b2f5">

**Comment by WhimsySpoon** [comment]
> I don't remember seeing this Public URL option before. Is it relevant? I've populated it but it hasn't changed anything, but I won't be in a position to remove and re-add the integration until later today.
> 
> ![image](https://user-images.githubusercontent.com/19993138/271548603-2a7a472c-045e-487c-a9bc-d172baae116f.png)

I removed, re-added and inserted my URL into the configuration and restarted. Unfortunately there was no change.

**Comment by popy2k14** [comment]
@all which are having issues with the new version and @all where it does work:

Which site do you using? in my case, i am using amazon.de

**Comment by WhimsySpoon** [comment]
I'm having issues with v4.7.3 using amazon.co.uk

**Comment by hukoeth** [comment]
Having issues with amazon.de.

**Comment by elarsson1** [comment]
Having issues, not seeing entities update, with 4.7.3 and amazon.com.

**Comment by amaisano** [comment]
Working perfectly (tested only 1x time on an Echo Show) "Set a timer / stop time" on `4.7.3`. I will update this comment if it stops working but so far so good.

I made no changes except installing the update and restarting Home Assistant.

Config data including region URLs:

```
var ue_id = 'AMXXXXXXXXXXX',
    ue_url = '/ap/uedata',
    ue_navtiming = 1,
    ue_mid = 'XXXXXXXX',
    ue_sid = '141-XXXXXXX-XXXXXX',
    ue_sn = 'www.amazon.com',
    ue_furl = 'fls-na.amazon.com',
    ue_surl = 'https://unagi-na.amazon.com/1/events/com.amazon.csm.nexusclient.prod',
    ue_int = 0,
    ue_fcsn = 1,
    ue_urt = 3,
    ue_rpl_ns = 'cel-rpl',
    ue_ddq = 1,
    ue_fpf = '//fls-na.amazon.com/1/batch/1/OP/ATVPDKIKX0DER:141-XXXXXXX-XXXXXX:XXXXXXXXX$uedata=s:',
    ue_sbuimp = 1,
    ue_ibft = 0,
    ue_fnt = 0,
```

**Comment by WhimsySpoon** [comment]
If you play some media, does it update too?

**Comment by popy2k14** [comment]
Thx at all, so it has nothing todo with a particular site.
amaisano: interessting that it works for you and for the most of us not.

Does it update when you play some media?

**Comment by amaisano** [comment]
Yes. It works flawlessly with Media controls. I can start it with voice, and pause it with HA, and resume it with voice. The HA status is delayed by 1-2 seconds, but it doesn't get much better than this.

![image](https://github.com/custom-components/alexa_media_player/assets/688435/09d2bddc-1170-46d3-9631-785779d65739)

**Comment by Camatobe** [comment]
> @ALL which are having issues with the new version and @ALL where it does work:
> Which site do you using? in my case, i am using amazon.de

Still the issue with media player entities. – using amazon.de
they are not updating at all. - in the alexa app everything is fine. 

Tried a lot of things: reinstalling the integration, restarting all echo devices, restarting HA, reloading integration. Cant even get the right media after reloading the integration anymore.

the log: no occurences of the alexa media player integration.

**Comment by rschultzbr** [comment]
I just installed the 4.7.3 update and restarted the home assistant server and everything started working perfectly!

**Comment by elarsson1** [comment]
After removing and re-adding the integration with 4.7.3, I am once again getting push messages, but the entites are only updating after a subsequent voice command. 

Example:
I say "Alexa, set an alarm for 7:30am"
Next alarm sensor shows "Unknown"
I say "Alexa, what time is it?"
Next alarm sensor shows "Saturday September 30 7:30am"
I say "Alexa, cancel alarm"
Next alarm sensor shows "Saturday September 30 7:30am"
I say "What is today's weather"
Next alarm sensor shows "Unknown""

I've repeated this with varying lengths of delay in between steps from 5 seconds to 10 minutes with the same result. I believe that some of the duplicate suppression logic introduced to overcome previous issues might be causing the problem here and will dig into that code.

**Comment by amaisano** [comment]
Update from https://github.com/custom-components/alexa_media_player/issues/1953#issuecomment-1741209351

Timers updated instantly both being set, snoozed, cancelled, and stopped on my Echo Show device. But my 5th gen "snowball" echo dot in the other room is still having problems.

I am not getting updates to timers in HA until I click my "ha.update_entity" service on it. I waited about 10 seconds to see if it was just taking longer. Note, this is an improvement still because the update_entity service is actually working, whereas before it still wouldn't see any changes...

But still, odd...

**Comment by serlinGi** [comment]
same issue for me: v4.7.3 , amazon.it

**Comment by amaisano** [comment]
> After removing and re-adding the integration with 4.7.3, I am once again getting push messages, but the entites are only updating after a subsequent voice command.
> 
> Example: I say "Alexa, set an alarm for 7:30am" Next alarm sensor shows "Unknown" I say "Alexa, what time is it?" Next alarm sensor shows "Saturday September 30 7:30am" I say "Alexa, cancel alarm" Next alarm sensor shows "Saturday September 30 7:30am" I say "What is today's weather" Next alarm sensor shows "Unknown""
> 
> I've repeated this with varying lengths of delay in between steps from 5 seconds to 10 minutes with the same result. I believe that some of the duplicate suppression logic introduced to overcome previous issues might be causing the problem here and will dig into that code.

I think that is what I'm experiencing as well, and agree. I forgot that I patched some of the code a while ago to and workaround this, but I believe those patches were destroyed when I installed 4.7.3. Sounds like some of them were added directly to the codebase though. Good luck - happy to test/confirm.

**Comment by elarsson1** [comment]
I fixed the issue with timers and alarms for me by partially revering https://github.com/custom-components/alexa_media_player/commit/8f0dfa2f6585ebac08bd5a708951155c80f63588

It seems that PUSH_NOTIFICATION_CHANGE events are implemented by Amazon over HTTP2. Restoring this bit of code has made my timers and alarms update within 5 seconds of a change on the device, and has the added bonus of responding to changes made on the Alexa app as well. I'll put together a small PR for the change.

**Comment by formatBCE** [comment]
4.7.3, amazon.ca

Media updates are good (sometimes lagging for up to 5 seconds, but stable).

Setting timer updated right away, but canceling did update only after another query to the echo.

Overall, I guess, it's enough for me.
Thanks for fixing!

**Comment by francescolf** [comment]
Apparently alexapy [hardcodes the us endpoint](https://gitlab.com/keatontaylor/alexapy/-/blob/dev/alexapy/alexahttp2.py#L53), hence eu instances don't receive push updates. Changing the authority to `bob-dispatch-prod-eu.amazon.com` fixed the issue.

**Comment by popy2k14** [comment]
@francescolf 
Nice find!
Didn't found alexapy in custom_components folder.

Where did you change the authority?

Thx

**Comment by francescolf** [comment]
> Didn't found alexapy in custom_components folder.

You should find it in your Python packages folder. The complete path should be `/usr/local/lib/python3.11/site-packages/alexapy/alexahttp2.py`

**Comment by WhimsySpoon** [comment]
Amazing find. Any way to change that on HAOS?

**Comment by francescolf** [comment]
> Amazing find. Any way to change that on HAOS?

I'm not very familiar with HAOS. If the path above doesn't work try one of [these](https://stackoverflow.com/questions/122327/how-do-i-find-the-location-of-my-python-site-packages-directory#:~:text=user%20site%2Dpackages.-,Practical%20Tips,-%3Cpackage%3E.__path__%20lets) methods.

**Comment by chrishas35** [comment]
4.7.3
amazon.com

Primary use case is to control Echo Glows which are "devices connected via Echo." Deleted and re-created the integration after upload, but these are still not updating. Clicking a Glow on in HA turns on the light, but HA pauses for a bit then the UI clicks back to off presumably because the service had not updated to on. The light remains on.

**Comment by KameDomotics** [comment]
> Apparently alexapy [hardcodes the us endpoint](https://gitlab.com/keatontaylor/alexapy/-/blob/dev/alexapy/alexahttp2.py#L53), hence eu instances don't receive push updates. Changing the authority to `bob-dispatch-prod-eu.amazon.com` fixed the issue.

Will be fixed with version 4.7.4?

**Comment by oneseventhree** [comment]
> Apparently alexapy [hardcodes the us endpoint](https://gitlab.com/keatontaylor/alexapy/-/blob/dev/alexapy/alexahttp2.py#L53), hence eu instances don't receive push updates. Changing the authority to `bob-dispatch-prod-eu.amazon.com` fixed the issue.

How about for Amazon.com.au?

**Comment by jeffalperin** [comment]
What is the appropriate way to update to 4.7.3?  I'm running HAOS and my version number seems to be 4.6.5 (discovered by inspecting `const.py`) even after deleting the integration, rebooting, re-adding the integration and rebooting.  Do I have to 1) wait for the next HA update, 2)manually replace all the files in the `/custom-components/alexa_media_player` folder or 3) something else?

**Comment by francescolf** [comment]
> How about for Amazon.com.au?

bob-dispatch-prod-fe.amazon.com

I'll try to patch alexapy to automatically choose the correct host.

**Comment by formatBCE** [comment]
> What is the appropriate way to update to 4.7.3?  I'm running HAOS and my version number seems to be 4.6.5 (discovered by inspecting `const.py`) even after deleting the integration, rebooting, re-adding the integration and rebooting.  Do I have to 1) wait for the next HA update, 2)manually replace all the files in the `/custom-components/alexa_media_player` folder or 3) something else?

I went to repo in HACS and in menu pushed "update info". After that it popped out.

**Comment by jeffalperin** [comment]
ugh.  I confused this with one of the built-in integrations.  HACS does indeed allow me to update.  Thanks!

**Comment by oneseventhree** [comment]
> ugh. I confused this with one of the built-in integrations. HACS does indeed allow me to update. Thanks!

redownload it from hacs on the right:
<img width="1394" alt="image" src="https://github.com/custom-components/alexa_media_player/assets/15043282/75979820-07b0-4be0-bd9e-9c4d66a44497">

**Comment by alandtse** [comment]
Please note, we'll do another release to address this but after that releases will slow down. We are missing many translations for some of the new keys that we were required to add and they only get added during a release. If you're in a non-English locale, it'd be helpful to get [those filled out](https://github.com/custom-components/alexa_media_player#translations) or you'll just see blanks where we don't have your language.

> > How about for Amazon.com.au?
> 
> bob-dispatch-prod-fe.amazon.com
> 
> I'll try to patch alexapy to automatically choose the correct host.

https://github.com/Apollon77/alexa-remote/blob/master/alexa-http2push.js#L33-L44 has the conversions.

**Comment by Camatobe** [comment]
> Please note, we'll do another release to address this but after that releases will slow down. We are missing many translations for some of the new keys that we were required to add and they only get added during a release. If you're in a non-English locale, it'd be helpful to get [those filled out](https://github.com/custom-components/alexa_media_player#translations) or you'll just see blanks where we don't have your language.
> 
> > > How about for Amazon.com.au?
> > 
> > 
> > bob-dispatch-prod-fe.amazon.com
> > I'll try to patch alexapy to automatically choose the correct host.
> 
> https://github.com/Apollon77/alexa-remote/blob/master/alexa-http2push.js#L33-L44 has the conversions.

Keep in mind, the conversions aren't complete! amazon.de (Germany) is missing for example. https://github.com/Apollon77/alexa-remote/blob/master/alexa-http2push.js#L33-L44

**Comment by conorlap** [comment]
> 
> Keep in mind, the conversions aren't complete! amazon.de (Germany) is missing for example. https://github.com/Apollon77/alexa-remote/blob/master/alexa-http2push.js#L33-L44

Amazon.co.uk is also missing!

**Comment by Profex** [comment]
amazon.de (or amazon.it, amazon.fr, amazon.uk etc.) is not missing, it's covered by the default selection
(https://github.com/Apollon77/alexa-remote/blob/master/alexa-http2push.js#L33)

**Comment by francescolf** [comment]
4.7.4 works as expected.

Thanks alandtse 👍

**Comment by popy2k14** [comment]
Also for me 4.7.4 works as expected
Thanks a lot alantse!!

**Comment by herveaurel** [comment]
4.7.4 only works 5 times and stops working. music states are not updated after 5 or 6 changes. 😫

**Comment by popy2k14** [comment]
You are right, it also stopped working for me now here, after some notifications.

**Comment by emufan** [comment]
Would be worth if you all test a little bit deeper and longer instead of returning, that it is working directly now and there.

**Comment by francescolf** [comment]
Also make sure you're using v4.7.5, which should fix this specific problem.

**Comment by WhimsySpoon** [comment]
Confirmed that 4.7.5 fixes the issue.

I noticed the same problem in 4.7.4 where eventually the HTTP/2 connection would close with:

`[alexapy.alexahttp2] HTTP2 Connection Closed.`

I've been running 4.7.5 for about 20 minutes and I've lost count of how many commands I sent and the connection has remained open.

Great work on this to all involved. Thank you.

**Comment by herveaurel** [comment]
4.7.5 : ❤️❤️❤️

**Comment by bcmitri** [comment]
I must be in the minority here. 4.7.3 worked mostly fine for me, but 4.7.4 and 4.7.5 do not work for me at all. I’m in Canada, and I honestly don’t remember what server I’m connected to, but something in 4.7.4 definitely broke things for me. I’ll have to do some more testing to see if I can pin the issue down.

**Comment by TheDeadEndLink** [comment]
Confirmed, 4.7.5 fixes the issue for me with Amazon.de.

A huge thank you to everyone who worked hard on this solution. This integration is the most valuable to me of all the HA integrations.

**Comment by oneseventhree** [comment]
Mine didn’t work but deleting the Integration in HA then adding it again and it worked. No restart needed :) thanks!

**Comment by serlinGi** [comment]
everything perfect 4.7.5, thank you

**Comment by alandtse** [comment]
@francescolf or anyone who can manually check: mind checking if this EU domain works for you? `alexa.eu.gateway.devices.a2z.com` I want to future proof this by switching domains.

https://developer.amazon.com/en-US/docs/alexa/alexa-voice-service/api-overview.html#endpoints

**Comment by francescolf** [comment]
Yes, works fine. They point to the same IP address.
![image](https://github.com/custom-components/alexa_media_player/assets/14892143/25f7421d-3398-4832-aa43-f04de0b47033)
(`alexa.eu.gateway.devices.a2z.com` is also an alias of `default.eu.gateway.devices.a2z.com`)

**Comment by chrishas35** [comment]
4.7.5
amazon.com

Primary use case of controlling Echo Glows which are "devices connected via Echo" still not working.  Have done delete, reinstall, restart between it all.  Behavior exactly same as before.

**Comment by smartbutnot** [comment]
Thanks so much, my timer display works again!

**Comment by KameDomotics** [comment]
After updating to version 4.7.6, today Alexa Media Player requested re-authentication. With previous versions (4.6.5) this had never happened.

I use Amazon EU (IT), Hass OS 2023.9.2.

Also, is so many API calls normal?

I attach screenshots of the relogin notifications:
![IMG_5872](https://github.com/custom-components/alexa_media_player/assets/47695661/e5ee9b8c-2258-449c-b99b-20ebd4802ec9)


![IMG_5882](https://github.com/custom-components/alexa_media_player/assets/47695661/e4a09dc6-74f0-493b-a7bf-6254b22373f0)

**Comment by apbarratt** [comment]
So I've been updating regularly, including the latest one today, and I've got no volume controls in the entities for all my media players.  They've been missing for probably a couple of months now.
I saw a closed ticket about this which has redirected me to this ticket, which seems also to be closed.  Am I in the right place?  Why is this ticket so very active and also marked as closed? :/

**Comment by sixdas** [comment]
I also no longer have volume controls, plus it always appears to be waiting.

I tried reinstalling several times but to no avail.

System details

Home-assistant (version): 2023.5.3
alexa_media (version from const.py or HA startup): 4.7.6

**Comment by serlinGi** [comment]
> After updating to version 4.7.6, today Alexa Media Player requested re-authentication. With previous versions (4.6.5) this had never happened.
> 
> I use Amazon EU (IT), Hass OS 2023.9.2.
> 
> Also, is so many API calls normal?
> 
> I attach screenshots of the relogin notifications: ![IMG_5872](https://user-images.githubusercontent.com/47695661/271856979-e5ee9b8c-2258-449c-b99b-20ebd4802ec9.jpeg)
> 
> ![IMG_5882](https://user-images.githubusercontent.com/47695661/271857005-e4a09dc6-74f0-493b-a7bf-6254b22373f0.jpeg)

same here '(

**Comment by tvizzotto** [comment]
Updated to v4.7.6 and everything stopped working.
Uninstalled, went back to v4.7.5 and everything bring back to life :)
PS: using amazon.com.br

**Comment by Pirol62** [comment]
Guys: useless to post here further in a closed thread.
The authentication issue is handled here: https://github.com/custom-components/alexa_media_player/issues/2059
and should have been solved with 4.7.7

**Comment by emufan** [comment]
But did you test 4.7.7? As you said you only reveted from 4.7.5 to 4.7.3.

**Comment by Pirol62** [comment]
Then the best way is, to open a new issue on GitHub an provide logs (simply raise the error level).
Maybe you should follow the way from tvizzotto first to aware that after installing you really use the installed version.
4.7.7 is working without any issues here and as the thread is closed, you might not get any real help here.

**Comment by bcmitri** [comment]
> Then the best way is, to open a new issue on GitHub an provide logs (simply raise the error level). Maybe you should follow the way from tvizzotto first to aware that after installing you really use the installed version. 4.7.7 is working without any issues here and as the thread is closed, you might not get any real help here.

I'm not sure what you mean by "really using the installed version".  I'd be happy to log a new bug, but I have no idea what I'm looking for in the logs.

**Comment by jaxfellow** [comment]
> I also no longer have volume controls, plus it always appears to be waiting.
> 
> I tried reinstalling several times but to no avail.
> 
> System details
> 
> Home-assistant (version): 2023.5.3 alexa_media (version from const.py or HA startup): 4.7.6

Sorry if I’m reviving this, I’m just newly installed the integration and searching for the issues before opening a new one.

I just installed it yesterday, custom commands and scripts running properly but I also don’t have any volume control.

**Comment by 1liminal1** [comment]
Yeah sadly Amazon have nutered a lot of this. The devs have done there upmost to have this functionality, but evilcorp just keeps shutting it off. We are luckily to have anything at all considering how much Amazon is hemorrhaging with Alexa. 

Im hoping that either HA's own voice stuff matures to the point we can just be rid of Alexa and let it die. Or that the paid Alexa tier will reenable this functionality

**Comment by alandtse** [comment]
Also volume is only available when select media is playing. Mainly Amazon music.

**Comment by jaxfellow** [comment]
> Also volume is only available when select media is playing. Mainly Amazon music.

I managed to control the volume using a script and buttons after all

**Comment by 1liminal1** [comment]
I found controlling media_player.spotify to be the most reliable, but it it works that's awesome :)
